### PR TITLE
[dev] Optimize MTP contiguous packed-CP rolls

### PR DESCRIPTION
**Pull request by hxbai**
Prefetch successor rows once per microbatch via a layout-neutral roll context so repeated MTP rolls avoid per-depth P2P.

- [ ] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?
Introduce a layout-neutral MTPSequenceRollContext so GPT/Hybrid prepare MTP sequence rolls once per microbatch, then reuse that context across MTP layers without layout-specific arguments.

For contiguous packed CP, prefetch compact successor “halo” rows for all present MTP fields (input_ids / position_ids / labels / loss_mask / padding_mask) in one grouped P2P exchange of width mtp_num_layers + 1, then make subsequent rolls local. Keep the old per-roll grouped P2P path as a correctness fallback (e.g. tiny shards or missing fields).

Make roll_tensor operate on tensor groups, vectorize the CP1 packed path (torch.roll + boundary index_fill_), and align Hybrid with GPT by resolving the effective CP group via resolve_cp_group.

Gate MTP acceptance-stat collection on whether TensorBoard/W&B will consume it (using globally consistent flags so ranks stay aligned), and pack correct/total counts into one reducible tensor.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/transformer/multi_token_prediction.py [hunk 1]
```diff
@@ -135,53 +534,97 @@ def tie_output_layer_state_dict(
...
-    and tensor rolling requires communication between adjacent CP ranks to properly handle the
-    boundary conditions.
-
-    For CP=1 (default behavior): Uses standard torch.roll with zero padding
-    For CP>1: Splits tensor into chunks, performs rolling within each chunk, then exchanges
-    boundary elements between adjacent CP ranks to maintain sequence continuity.
-
-    For packed sequences: Respects sequence boundaries when rolling to avoid mixing tokens
-    from different sequences.
+def roll_tensor(
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/multi_token_prediction.py:537]: It is OK that no other downstream places use it.

File: megatron/core/transformer/multi_token_prediction.py [hunk 2]
```diff
@@ -57,6 +57,405 @@
...
+        padding_mask: Successor padding flags, with True marking padding.
+    """
+
+    input_ids: Optional[Tensor] = None
+    position_ids: Optional[Tensor] = None
+    labels: Optional[Tensor] = None
+    loss_mask: Optional[Tensor] = None
+    padding_mask: Optional[Tensor] = None
+
+    def __post_init__(self):
```
[the remark this answers is not in the document]
Comment (hxbai) [reply megatron/core/transformer/multi_token_prediction.py:107]: I'll leave it as it is now. Adding an abstraction is more complicated.

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/6246/commits/db4fe4609b87a0e7290a7922fee911e2d49f8905

**Review by yuzhongw-nvidia — APPROVED** [review]
LGTM. Thanks.

**Review by yaox12 — APPROVED** [review]

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/6246/commits/78f7968abbc44f419f5ae23ceea774ce8c8adc32
